feat: 로그인 페이지에 로직 추가

smb0123 opened the pull request:
### 개요

close #6

### 작업사항

- 카카오 로그인 로직을 추가하였습니다.
- 카카오 로그인 후 리다이렉트될 페이지를 구현하였습니다. (social-login-loading)
- tanstack query를 적용하였습니다.

### 변경로직

- 카카오, 애플 로그인 버튼을 컴포넌트화 시켰습니다.
- Flex 컴포넌트에 height props를 추가하였습니다.

### 하고 싶은 말
이전에 올린 pr에서 develop 브랜치를 합치는 과정에서 문제가 생겨서 새롭게 브랜치를 만들어서 작업하였습니다.
코드는 이전 pr에서 피드백 받은 부분을 반영해서 그 부분만 보면 될 것 같습니다.

### 체크리스트

- [x] assignees 설정
- [x] label 설정

smb0123 committed:
feat: api 통신을 위한 instance 구현

instance와 interceptors를 구현하였습니다.

smb0123 committed:
feat: 카카오 로그인 api 함수 구현

카카오 로그인을 위한 api 함수입니다.

smb0123 committed:
chore: axios 라이브러리 설치

api 통신을 위해 설치하였습니다.

smb0123 committed:
feat: TanstackQueryProvider 컴포넌트 구현

tanstack query 라이브러리 설치 후 이를 적용한 컴포넌트를 구현하여
전역에서 tanstack query를 사용할 수 있게 하였습니다.

smb0123 committed:
feat: 경로 상수 구현

경로를 모아 놓은 상수를 만들었습니다.

smb0123 committed:
refactor: Flex 컴포넌트에 height props 추가

smb0123 committed:
feat: 소셜 로그인 로딩 페이지 구현

소셜 로그인 후 이를 처리할 페이지를 구현하였습니다.

smb0123 committed:
refactor: 카카오 로그인 버튼과 애플 로그인 버튼을 컴포넌트로 분리

유지보수를 위해서 로그인 버튼을 컴포넌트로 분리하였고 카카오 로그인 훅을
구현하였습니다.

smb0123 committed:
refactor: LoginTitleSection에 section 태그 추가

geongyu09 commented on src/components/features/pages/LoginButtonSection/index.tsx:
```diff
@@ -1,49 +1,15 @@
-import Button from "@shared/ui/Button";
 import Spacing from "@shared/layout/Spacing";
+import AppleLoginButton from "../../widgets/AppleLoginButton";
+import KakaoLoginButton from "../../widgets/KakaoLoginButton";
```
alias로 변경해도 좋을 것 같습니다!

geongyu09 commented on src/app/social-login-loading/page.tsx:
```diff
@@ -0,0 +1,10 @@
+export default function SocialLoginLoadingPage() {
+  return (
+    <Suspense>
+      <SpinnerSection />
```
일단 로딩 관련 처리는 디자이너 및 기획자와 협의된게 없어서 일단 제거하는게 맞을 것 같습니다.
이는 논의 후 추후에 알려드리도록 하겠습니다

geongyu09 commented on src/app/social-login-loading/page.tsx:
```diff
@@ -0,0 +1,10 @@
+
+export default function SocialLoginLoadingPage() {
+  return (
+    <Suspense>
```
이런식으로 hydration 에러를 해결할 수도 있군요 👍👍

geongyu09 commented on src/components/common/shared/layout/Flex/index.tsx:
```diff
@@ -38,14 +39,22 @@ export default function Flex({
+
+  if (height) {
+    heightClass = typeof height === "number" ? `h-${height}` : height;
+  }
```
개인적으로 명령적인 로직 보다는 선언적으로 작성하는게 더 좋을 것 같습니다

```
const heightClass = height 
  ? (typeof height === "number" ? `h-${height}` : height)
  : undefined;
```

다른 분들은 어떤 것 같나요??

seunghyeon77 replied:
건규님은 명령적인 로직(일반 조건문)과 선언적인 로직(삼항 연산자) 간의 성능적인 차이가 크다기 보다는, 명시적으로 다가오는 차이 때문에 선언적인 로직을 더 선호하시는 건가요??

저 또한 선언적인 로직(삼항 연산자)가 더 좋다고 생각합니다! 왜냐하면, 따로 더 많은 경우를 분기 처리하지 않고, 변수에 값을 할당하는 것이 더 자연스럽기 때문입니다!

geongyu09 replied:
명령형, 선언형은 패러다임의 차이이지 성능의 차이가 아닙니다!
리액트 코드는 애초에 선언형 로직에 맞추어서 개발되었고, 사용할 때에도 선언적으로 작성하도록 유도한 것들이 많습니다. 
선언적으로 작성하면 가독성이 좋아지고, 동작이 추상화 되어 개발시 결과물에만 집중할 수 있습니다. 
물론 이것이 매 순간에 적용되는 사항은 아니라고 생각합니다. 
그래서 해당 코드를 제가 제안한 방식으로 변경하는것이 괜찮은지 여쭤본 것이었습니다

smb0123 replied:
저는 삼항 연산자를 2번 사용하는 게 가독성에 좋지 않다고 생각해서 명령형으로 작성을 하였는데 다른 더 좋은 방법에 대해 생각해보겠습니다.

geongyu09 commented on src/apis/instances/index.ts:
```diff
@@ -0,0 +1,3 @@
+import instance from "@/apis/instances/instance";
```
api의 path를 지정할 때 폴더 구조로 URI를 관리하기로 했었는데, instances가 들어가 있으면 분명히 헷갈릴 수 있기도 하고, 나중에 해당 URI가 생기게 되면 문제가 발생할 수 있을 것 같습니다
next에서 app 폴더를 통해서 라우팅을 지정할 때 `_` 를 prefix로 붙이게 되면 라우팅에 잡히지 않는 것을 알고 계신가요?
이처럼 저희도 _instances를 사용하는건 어떠신가요?

추가적으로 instance.ts 파일 이름이 조금 거슬리는 것 같기도 합니다..! 다른 코드들은 모두 index인데 이것만(훅도 그러긴 함) 이름을 가지는 것도 뭔가 이상하기도 합니다..!
어쩌면 말 바꾸기 이기도 하지만, 한번 다시 논의할 필요도 있는 것 같습니다

seunghyeon77 replied:
저도 _instances로 폴더 이름을 명명하는 것이 좋을 거 같습니다! 해당 이름의 API가 생겼을 경우, Next에서 잘못 라우팅이 이루어지는 것을 방지할 수 있을 거 같네요!

저는 다른 부분과의 통일성을 위해, index.ts로 맞추는 게 좋을 거 같아요! 그럼 instances 폴더 아래 index.ts에 re-import한 파일들을 모아넣고, instance.ts 파일은 instance 폴더 아래 index.ts로 두는 걸 생각했습니다!

geongyu09 replied:
여기서 혼돈해서는 안되는게 애초에 /apis에 있는 로직은 next에서 라우팅을 지정하지않습니다!

언급했던 훅과 api instance 부분에 대해서는 조금 더 고민해보고 이야기 해보면 좋을 것 같습니다

smb0123 replied:
> 추가적으로 instance.ts 파일 이름이 조금 거슬리는 것 같기도 합니다..! 다른 코드들은 모두 index인데 이것만(훅도 그러긴 함) 이름을 가지는 것도 뭔가 이상하기도 합니다..!

저도 동의합니다. 하지만 app 폴더에 넣는 것은 애매하네요.

geongyu09 replied:
앗 app 폴더에 넣자 라는게 아니라, app 폴더에서 라우팅을 하는 폴더와 라우팅을 하지 않는 폴더를 _ 로 구분한다는 것을 설명한 거였어요!
단순 예시이고, apis 에 있으니 당연히 next의 라우팅에는 안잡힌다라는 말이었습니다

"app 폴더가 이런식으로 동작하는데, 우리도 apis에 이런거 도입해보는건 어때?" 정도롤 이해해주세요!

smb0123 replied:
아하 그런 뜻이었군요...!
괜찮은 방법인 것 같습니다.

geongyu09 reviewed (changes requested):
수고 많으셨습니다
아마 다음에 개발할 내용은 앱과의 통신 로직이 되겠네요(맞나요??)
혹시나 해당 내용 개발을 빠르게 진행해야한다면 제가 작업하도록 하겠습니다

일단 로딩 화면에 대해서는, 우선적으로 빼는것이 맞을 것 같습니다. 
먼저 논의해보고 말씀 드릴게요!

추가적으로, 훅과 관련해서도 지금의 방법이 썩 좋아보이지는 않는 느낌을 받습니다. 
굳이 뭔가 힘들게 도메인으로 나눈 느낌도 들고, 다른 코드들과 다르게 파일 이름이 index가 아니기도 하고요..
이는 한번 이야기 해보면 좋을 것 같습니다.

seunghyeon77 reviewed (commented):
수고 많으셨습니다!

smb0123 committed:
refactor: 버튼 컴포넌트에 ...props 추가

버튼 태그가 가질 수 있는 속성을 전부 props로 받을 수 있도록
수정하였습니다.

smb0123 committed:
chore: instance의 baseURL 변경

smb0123 committed:
chore: postKakaoLogin 함수 폴더 위치 변경

api 경로에 맞게 폴더 구조를 변경하였습니다.

smb0123 committed:
chore: import문을 alias로 변경

smb0123 committed:
chore: social-login-loading 페이지에서 스페너 주석 처리

smb0123 committed:
refactor: HeightType 정의 후 heightClass 제거

heightClass는 코드가 깔끔하지 않아서 HeightType를 정의하고 이를
제거하였습니다.

smb0123 committed:
chore: 프레그먼트 태그 삭제

geongyu09 commented on src/type/css/height.ts:
```diff
@@ -0,0 +1,41 @@
+export type HeightType =
```
결국 금단의 코드가 나오게 되었네요..! 
(tailwind.... 타입까지 지원하면 얼마나 좋을까요?)

해당 타입은 height, width에 동일하게 적용이 가능합니다! 
조금 더 범용적으로(width, height) 사용이 가능할 것 같은데, 이름을 SizeType 등으로 변경하는것은 어떤가요?
(이름은 그대로 따를 필요 없습니다!)

geongyu09 replied:
아니면, 숫자는 gap과 같은 곳에도 사용이 되니까 따로 타입으로 빼도 괜찮을... 것 같기도 하네요..! 😅

geongyu09 reviewed (approved):
굳굳 좋습니다!!

smb0123 committed:
chore: instances 폴더명 _instances로 변경

smb0123 committed:
chore: Flex의 변경된 height props로 인한 변경

smb0123 committed:
refactor: 타입명 변경 및 숫자 사이즈 값만 가지는 타입 추가

width에서도 사용할 수 있도록 타입명을 SizeType로 변경하였고 gap에서
사용할 수 있도록 SizeNumberType도 구현하였습니다.

smb0123 committed:
refactor: gap의 타입으로 SizeNumberType 사용

global.css의 gap에 대한 내용을 변경하고 SizeNumberType을 적용하였습니다.

smb0123 committed:
chore: 생략된 } 추가

smb0123 committed:
refactor: Spacing 컴포넌트의 size props에 SizeNumberType 적용

smb0123 reviewed (commented):
타입을 분리하고 gap과 mt에도 적용하였습니다.

geongyu09 reviewed (approved):
좋아용!! 완벽합니다
